subsys/bluetooth/mesh/delayable_msg.c

```diff
@@ -0,0 +1,278 @@
+	}
+
+	return number;
+}
+
+static void release_delayable_msg_chunks(struct delayable_msg_ctx *msg)
+{
+	struct delayable_msg_chunk *chunk;
+	sys_snode_t *node;
+
+	while ((node = sys_slist_get(&msg->chunks))) {
+		chunk = CONTAINER_OF(node, struct delayable_msg_chunk, node);
+		k_mem_slab_free(&delayable_msg_chunks, (void *)chunk);
+	}
+}
+
+static struct delayable_msg_ctx *allocate_delayable_msg_ctx(void)
+{
+	struct delayable_msg_ctx *msg = NULL;
+
+	if (k_mem_slab_alloc(&delayable_msgs_ctx, (void **)&msg, K_NO_WAIT)) {
+		LOG_WRN("Purge pending delayable message.");
+		if (push_msg_from_delayable_msgs()) {
+			if (k_mem_slab_alloc(&delayable_msgs_ctx, (void **)&msg, K_NO_WAIT)) {
+				LOG_WRN("No available delayable access message context.");
+			}
+		}
+	}
+
+	if (msg) {
+		sys_slist_init(&msg->chunks);
+	}
+
+	return msg;
+}
+
+static void release_delayable_msg_ctx(struct delayable_msg_ctx *ctx)
+{
+	(void)sys_slist_find_and_remove(&access_delayable_msg.busy_ctx, &ctx->node);
+	k_mem_slab_free(&delayable_msgs_ctx, (void *)ctx);
+}
+
+static bool push_msg_from_delayable_msgs(void)
+{
+	sys_snode_t *node;
+	struct delayable_msg_chunk *chunk;
+	struct delayable_msg_ctx *msg = get_pending_msg();
+	size_t len = msg->len;
+	int err;
+
+	if (!msg) {
+		return false;
+	}
+
+	NET_BUF_SIMPLE_DEFINE(buf, BT_MESH_TX_SDU_MAX);
+
+	SYS_SLIST_FOR_EACH_NODE(&msg->chunks, node) {
+		size_t tmp = MIN(CONFIG_BT_MESH_ACCESS_DELAYABLE_MSG_CHUNK_SIZE, len);
+
+		chunk = CONTAINER_OF(node, struct delayable_msg_chunk, node);
+		memcpy(net_buf_simple_add(&buf, tmp), chunk->data, tmp);
+		len -= tmp;
+	}
+
+	msg->ctx.rnd_delay = false;
+	err = bt_mesh_access_send(&msg->ctx, &buf, msg->src_addr, msg->cb, msg->cb_data);
+	msg->ctx.rnd_delay = true;
+	if (err != -EBUSY && err != -ENOBUFS) {
+		release_delayable_msg_chunks(msg);
+		release_delayable_msg_ctx(msg);
+		if (err && msg->cb && msg->cb->start) {
+			msg->cb->start(0, err, msg->cb_data);
+		}
+		return true;
+	}
+
+	return false;
+}
+
+static void delayable_msg_handler(struct k_work *w)
+{
+	if (!push_msg_from_delayable_msgs()) {
+		sys_snode_t *node = sys_slist_get(&access_delayable_msg.busy_ctx);
+		struct delayable_msg_ctx *pending_msg =
+			CONTAINER_OF(node, struct delayable_msg_ctx, node);
+
+		pending_msg->fired_time += 10;
+		reschedule_delayable_msg(pending_msg);
+	} else {
+		reschedule_delayable_msg(NULL);
+	}
+}
+
+int bt_mesh_access_delay(struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf, uint16_t src_addr,
+			 const struct bt_mesh_send_cb *cb, void *cb_data)
+{
+	sys_snode_t *node;
+	struct delayable_msg_ctx *msg;
+	uint16_t random_delay;
+	int total_number = DIV_ROUND_UP(buf->size, CONFIG_BT_MESH_ACCESS_DELAYABLE_MSG_CHUNK_SIZE);
+	int allocated_number = 0;
+	size_t len = buf->len;
+
+	if (total_number > CONFIG_BT_MESH_ACCESS_DELAYABLE_MSG_CHUNK_COUNT) {
+		return -EINVAL;
+	}
+
+	msg = allocate_delayable_msg_ctx();
+	if (!msg) {
+		return -ENOMEM;
```

Comment:
I think warnings will be useful here for 2 these error cases to minimize debugging.